### PR TITLE
SI-6187 Make partial functions re-typable

### PR DESCRIPTION
**Pull request by retronym**
- `New(tpe)` doesn't survive a `resetAttrs` / typecheck; use
  a name instead.
- Abandon the tree attachment that passed the default
  case from `typer` to `patmat`; this tree eluded the
  attribute reset performed in the macro. Instead, add it
  to the match. Apart from making the tree re-typable, it
  also exposes the true code structure to macros, which
  is important if they need to perform other code
  transformations.
- Install original trees on the declared types of the
  parameters of the `applyOrElse` method to ensure that
  references to them within the method pick up the correct
  type parameter skolems upon retypechecking.
- Propagate `TypeTree#original` through `copyAttrs`,
  which is called during tree duplication / `TreeCopiers`.
  Without this, the original trees that we installed were
  not visible anymore during `ResetAttrs`.

We are not able to reify partial functions yet -- the
particular sticking point is reification of the parentage
which is only available in the `ClassInfoType`.

Review by @adriaanm / @xeno-by 

I'm using `typedCase` now, based on your prior review. The `asInstanceOf` is also gone.

### CODE REVIEW

**Comment by retronym** [comment]
Sigh...

**Comment by retronym** [comment]
Review by @xeno-by @adriaanm

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -2706,7 +2709,7 @@ trait Typers extends Modes with Adaptations with Tags {
         val x = mkParam(methodSym, A1.tpe)
 
         // applyOrElse's default parameter:
-        val B1 = methodSym newTypeParameter (newTypeName("B1")) setInfo TypeBounds.empty //lower(resTp)
+        val B1 = methodSym newTypeParameter (newTypeName("B1")) setInfo TypeBounds.empty
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:2712]: `lower(resTp)`?

Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:2712]: Looking at this again, I see that comment was succinctly noting that this type will be patched a few lines below.

``` scala
         B1 setInfo TypeBounds.lower(matchResTp) // patch info
```

File: src/reflect/scala/reflect/internal/Trees.scala [hunk 2]
```diff
@@ -516,7 +516,11 @@ trait Trees extends api.Trees { self: SymbolTable =>
     override private[scala] def copyAttrs(tree: Tree) = {
       super.copyAttrs(tree)
       tree match {
-        case other: TypeTree => wasEmpty = other.wasEmpty // SI-6648 Critical for correct operation of `resetAttrs`.
+        case other: TypeTree =>
+          // SI-6648 Critical for correct operation of `resetAttrs`.
+          wasEmpty = other.wasEmpty
+          if (other.orig != null)
+            orig = other.orig.duplicate
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Trees.scala:523]: Wow! I wonder how come this wasn't here before :O

**Comment by adriaanm** [comment]
BUILDLOG?

**Comment by JamesIry** [comment]
@adriaanm and/or @xeno-by I think this PR is still waiting on review approval

**Comment by xeno-by** [comment]
Unfortunately, I know close to nothing about patmat, so I don't feel entitled for rubberstamping here

**Comment by adriaanm** [comment]
rebasing and reviewing

**Pull request closed without merging**